File: __tests__/features/payments/receiptScreenData.test.ts
import { projectReceipt } from '@/features/payments/ui/receiptScreenData';
import { emptyPayments, adding, type Payment } from '@/features/payments/data/payment';
import { emptySavedBillers, addingSaved } from '@/features/saved_billers/data/savedBiller';

const mk = (status: Payment['status']): Payment => ({
  id: 'pay-1', billerId: 'electricity-metro', billerName: 'Metro Electricity',
  categoryId: 'electricity', account: 'K123', amountPaise: 45000,
  paidAtUtc: new Date('2026-06-05T10:00:00.000Z'), status,
});

describe('projectReceipt', () => {
  it('unknown id is notFound', () => {
    expect(projectReceipt(emptyPayments(), emptySavedBillers(), 'pay-99')).toEqual({ kind: 'notFound' });
  });
  it('processing projects amount and biller', () => {
    const r = projectReceipt(adding(emptyPayments(), mk('processing')), emptySavedBillers(), 'pay-1');
    expect(r.kind).toBe('processing');
  });
  it('success offers save when biller not saved, hides when saved', () => {
    const payments = adding(emptyPayments(), mk('success'));
    expect(projectReceipt(payments, emptySavedBillers(), 'pay-1').kind === 'success' &&
      (projectReceipt(payments, emptySavedBillers(), 'pay-1') as { canSaveBiller: boolean }).canSaveBiller).toBe(true);
    const saved = addingSaved(emptySavedBillers(), { billerId: 'electricity-metro', account: 'K123', nickname: 'Home' });
    const r = projectReceipt(payments, saved, 'pay-1');
    expect(r.kind === 'success' && r.canSaveBiller).toBe(false);
  });
  it('failed projects retry fields', () => {
    const r = projectReceipt(adding(emptyPayments(), mk('failed')), emptySavedBillers(), 'pay-1');
    expect(r.kind === 'failed' && r.billerId).toBe('electricity-metro');
    expect(r.kind === 'failed' && r.categoryId).toBe('electricity');
  });
});

File: test/features/payments/receipt_screen_data_test.dart
import 'package:billpayments/core/mock/mock_network.dart';
import 'package:billpayments/features/auth/data/auth_provider.dart';
import 'package:billpayments/features/payments/data/payments_provider.dart';
import 'package:billpayments/features/payments/ui/receipt_screen_data.dart';
import 'package:billpayments/features/payments/ui/receipt_screen_data_provider.dart';
import 'package:billpayments/features/saved_billers/data/saved_biller.dart';
import 'package:billpayments/features/saved_billers/data/saved_billers_provider.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

ProviderContainer containerFailingEvery(int failEvery) {
  final container = ProviderContainer(overrides: [
    mockNetworkProvider.overrideWithValue(
        MockNetwork(minDelayMs: 0, maxDelayMs: 1, failEvery: failEvery)),
  ]);
  addTearDown(container.dispose);
  return container;
}

Future<void> login(ProviderContainer container) async {
  final notifier = container.read(authProvider.notifier);
  await notifier.sendOtp('9876543210');
  await notifier.verifyOtp('123456');
}

Future<void> payOnce(ProviderContainer container) =>
    container.read(paymentsProvider.notifier).pay(
        billerId: 'electricity-metro',
        billerName: 'Metro Electricity',
        categoryId: 'electricity',
        account: 'K123',
        amountPaise: 45000);

void main() {
  test('unknown payment id projects notFound', () async {
    final container = containerFailingEvery(100);
    await login(container);
    expect(container.read(receiptScreenDataProvider('pay-99')),
        const ReceiptScreenData.notFound());
  });

  test('successful payment projects receipt offering save biller', () async {
    final container = containerFailingEvery(100);
    await login(container);
    await payOnce(container);
    final data =
        container.read(receiptScreenDataProvider('pay-1')) as ReceiptSuccess;
    expect(data.billerName, 'Metro Electricity');
    expect(data.amountPaise, 45000);
    expect(data.canSaveBiller, isTrue);
  });

  test('already saved biller hides save action', () async {
    final container = containerFailingEvery(100);
    await login(container);
    container.read(savedBillersProvider.notifier).save(const SavedBiller(
        billerId: 'electricity-metro', account: 'K123', nickname: 'Home'));
    await payOnce(container);
    final data =
        container.read(receiptScreenDataProvider('pay-1')) as ReceiptSuccess;
    expect(data.canSaveBiller, isFalse);
  });

  test('declined payment projects failed with retry fields', () async {
    final container = containerFailingEvery(1);
    await login(container);
    await payOnce(container);
    final data =
        container.read(receiptScreenDataProvider('pay-1')) as ReceiptFailed;
    expect(data.billerId, 'electricity-metro');
    expect(data.amountPaise, 45000);
  });
}
